runtime/src/iree/hal/drivers/cuda/tracing.c

```diff
@@ -169,98 +227,189 @@ void iree_hal_cuda_tracing_context_free(
   if (!context) return;
-  if (context->query_tail == context->query_head) {
-    // No outstanding queries.
+  iree_slim_mutex_lock(&context->event_mutex);
+
+  // No outstanding queries
+  if (!context->submitted_event_list.head) {
+    iree_slim_mutex_unlock(&context->event_mutex);
     return;
   }
   IREE_TRACE_ZONE_BEGIN(z0);
-
-  while (context->query_tail != context->query_head) {
-    // Compute the contiguous range of queries ready to be read.
-    // If the ringbuffer wraps around we'll handle that in the next loop.
-    uint32_t try_query_count =
-        context->query_head < context->query_tail
-            ? context->query_capacity - context->query_tail
-            : context->query_head - context->query_tail;
-    IREE_TRACE_ZONE_APPEND_VALUE_I64(z0, (int64_t)try_query_count);
-
-    // Scan and feed the times to tracy, stopping when we hit the first
-    // unavailable query.
-    uint32_t query_base = context->query_tail;
-    uint32_t read_query_count = 0;
-    for (uint32_t i = 0; i < try_query_count; ++i) {
-      // Ensure the event has completed; will return CUDA_ERROR_NOT_READY if
-      // recorded but not retired or any other deferred error.
-      uint16_t query_id = (uint16_t)(query_base + i);
-      CUevent query_event = context->event_pool[query_id];
-      CUresult result = context->symbols->cuEventQuery(query_event);
+  // submitted_event_list is a list of the head elements for each
+  // command_buffer that has been submitted. Here we loop over
+  // all of the events, wait for them to complete and gather the results
+  // with cuEventQuery.
+
+  iree_hal_cuda_tracing_context_event_t* events =
+      context->submitted_event_list.head;
+  uint32_t read_query_count = 0;
+  // Outer per-command_buffer loop.
+  while (events) {
+    iree_hal_cuda_tracing_context_event_t* event = events;
+    // Inner per-event loop.
+    while (event) {
+      uint32_t query_id = (uint32_t)(event - &context->event_pool[0]);
+
+      CUresult result = context->symbols->cuEventSynchronize(event->event);
+      if (result != CUDA_SUCCESS) break;
+      result = context->symbols->cuEventQuery(event->event);
       if (result != CUDA_SUCCESS) break;
 
       // Calculate context-relative time and notify tracy.
       float relative_millis = 0.0f;
       IREE_CUDA_IGNORE_ERROR(
           context->symbols,
           cuEventElapsedTime(&relative_millis, context->base_event,
-                             query_event));
+                             event->event));
       int64_t gpu_timestamp = (int64_t)((double)relative_millis * 1000000.0);
       iree_tracing_gpu_zone_notify(context->id, query_id, gpu_timestamp);
 
-      read_query_count = i + 1;
+      read_query_count += 1;
+      event = event->next_in_command_buffer;
     }
-    IREE_TRACE_ZONE_APPEND_VALUE_I64(z0, (int64_t)read_query_count);
+    iree_hal_cuda_tracing_context_event_t* next = events->next_submission;
+    events->next_submission = events;
```

Comment:
Not sure I follow exactly what this is doing. This is creating a loop pointing to itself? Can you add some comments explaining it?